### PR TITLE
feat(grovedb,merk): provable offset on ProvableCountTree / ProvableCountSumTree single-range queries

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Summary

Adds a new proof flow that honors `SizedQuery::offset` for single-range queries against `ProvableCountTree` and `ProvableCountSumTree`. Skipped in-range subtrees collapse to a single hash-bound `HashWithCount` op (the same shape `AggregateCountOnRange` already uses), so the offset region pays **O(log n)** proof size per skipped subtree rather than O(skipped). Items inside the limit window emit as normal count-bearing value nodes, so the verifier-side result shape matches what a regular range query without offset would produce.

Closes the long-standing "no offsets on provable queries" gap for the count-tree case.

## What's reused vs. new

**Reuses:**
- `Node::HashWithCount` (existing opcode — no new proof variants).
- `node_hash_with_count` works for `ProvableCountSumTree` too because that variant's sum is intentionally not bound to the node hash, so a count-only `HashWithCount` is sufficient.
- The shared `aggregate_common::classify_subtree` Disjoint / Contained / Boundary classifier.

**New:**
- Module `merk/src/proofs/query/count_offset/{mod,prove,emit,verify,tests}.rs`.
- Public API: `Merk::prove_count_offset_on_range(range, offset, limit, left_to_right, ...)` and `verify_count_offset_on_range_proof`.
- Validators: `SizedQuery::validate_count_offset_paginated` + `PathQuery::validate_count_offset_paginated` + `PathQuery::has_non_zero_offset`.

## Scope

- **Tree types**: `ProvableCountTree` and `ProvableCountSumTree` only.
- **Query shape**: a single `QueryItem` range. Multi-item queries, subqueries, and conditional branches continue to reject offset (out of scope for this PR).
- **Direction**: both ascending and descending — the descending walk is a structural mirror (walks right child first, emits inverted ops, "first N in-range keys" = N highest keys).
- **Truncated offset**: when the requested offset exceeds the in-range population, the prover skips everything it can and returns 0 items. The verifier exposes this as `skipped < requested_offset` in the returned struct.

## Algorithm sketch

**Prover** (`emit_count_offset_proof`):
1. Classify the current subtree.
2. **Disjoint** → emit one `HashWithCount(count)` and bubble the structural count up. No offset/limit consumption.
3. **Contained** ∧ `subtree_count ≤ offset_remaining` → collapse to one `HashWithCount`, decrement offset.
4. **Contained** ∧ offset=0 ∧ limit exhausted → collapse to one `HashWithCount` (past-limit, no state change).
5. Otherwise descend per-element in directional order — emit `KVDigestCount` for path / skipped / past-limit positions, or a value-bearing node for returned items.

**Verifier**:
- Phase 1: reconstruct the proof tree via `execute_with_options`, allowlisting only the four node kinds an honest prover ever emits.
- Phase 2: walk the reconstructed tree directionally. `own_count` is derived in O(1) from each node's immediate children's count fields, so the in-order state machine knows the disposition *before* recursing into the second-direction child. The recursive return is then validated against the children's claimed count fields — locking the structural counts across the whole tree.
- State mutations are gated by `(in_range, own_count, offset_remaining, limit_remaining)`. Malformed shapes (digest at offset=0 with limit slots free, value emission with offset slots remaining, etc.) are rejected with precise error messages.

## GroveDB-layer wiring

- `prove_query_non_serialized_v{0,1}`: the hard offset rejection is replaced with a two-step gate — syntactic (`validate_count_offset_paginated`) + a tree-type check that opens the target merk. Both surface clear errors on mismatch.
- `prove_subqueries{,_v1}`: leaf-level short-circuit mirroring the existing aggregate-count / aggregate-sum branches.
- `verify_query_with_options`: same syntactic relaxation as the prover.
- `verify_layer_proof{,_v1}`: leaf-level dispatch routing to `verify_count_offset_on_range_proof`, then translating returned items into `ProvedPathKeyOptionalValue` rows the rest of the verifier expects.

## Tests

17 new tests, all green:

- **Merk-level** (`merk/src/proofs/query/count_offset/tests.rs`, 9 tests): round-trip on the 15-key fixture for offset+limit composition, both directions, empty trees, offset-past-end (truncated skip), partial-range queries, and tree-type rejection.
- **Grovedb-level** (`grovedb/src/tests/count_offset_paginated_tests.rs`, 8 tests): end-to-end through the full path-query stack including `ProvableCountSumTree`, descending direction, partial ranges, and the no-count-tree-target rejection case.

Two existing tests in `proof_coverage_tests.rs` were not modified — they assert offset on a NormalTree errors, and they still error (the new top-level tree-type gate fires before any descent).

Full suite status:
- merk: 502/502 ✅
- grovedb: 1721/1721 ✅

## Versioning

No new GroveVersion needed — this is purely additive to the proof flow at the latest version. The `HashWithCount` opcode and verifier shape were already there; this PR just extends the merk's set of code paths that emit them.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Offset and limit pagination support for count-tree based queries, enabling paginated retrieval of results.
  * New validation methods ensure pagination parameters meet eligibility requirements.

* **Tests**
  * Expanded test coverage for offset-paginated proofs, including validation scenarios, edge cases, and soundness checks across versions.

[Review Change Stack](https://app.coderabbit.ai/change-stack/dashpay/grovedb/pull/669?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by QuantumExplorer** [comment]
This is Claude. Pushed [`cc0ece07`](https://github.com/dashpay/grovedb/commit/cc0ece07) to address the codecov/patch coverage gap (was 67.4% in the first push).

Added 18 tests targeting the rejection branches the happy-path round-trips don't exercise:

**Merk-level adversarial verifier tests** (8):
- Verifier called with mismatched range / direction / offset (smaller and larger) / limit — each replays a state-machine inconsistency that should be rejected.
- Single-byte mutation, truncation, and trailing-garbage proofs — exercises the decoder + `execute_with_options` stack-size check + final hash divergence.

**GroveDB-level validator unit tests** (10):
- One per rejection branch in `SizedQuery::validate_count_offset_paginated` (no offset, offset=0, aggregate-count wrapper, aggregate-sum wrapper, default subquery, default subquery_path, multi-item).
- Sanity sweep that all 10 allowed `QueryItem` variants pass validation.
- `PathQuery`-level empty-path rejection + `has_non_zero_offset` truth table.

Test totals:
- Count-offset specific: **35 tests** (17 merk + 18 grovedb), all green.
- Full suites: merk 510/510, grovedb 1731/1731.

CodeRabbit's review status came back as "skipped" with only a walkthrough summary posted — no actionable inline comments to address. If you'd like a more thorough review, I can run `/ultrareview` on this PR (note: that's a user-triggered command, I can't launch it myself).

File: grovedb/src/operations/proof/generate.rs [hunk 1]
```diff
@@ -181,6 +181,50 @@ impl GroveDb {
...
+        );
+        if !matches!(
+            target.tree_type,
+            MerkTreeType::ProvableCountTree | MerkTreeType::ProvableCountSumTree
+        ) {
+            return Err(Error::InvalidQuery(
+                "count-offset paginated queries are only valid against \
+                 ProvableCountTree / ProvableCountSumTree merks",
+            ))
+            .wrap_with_cost(cost);
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply grovedb/src/operations/proof/generate.rs]: This is Claude. Fixed in b3f203d0 — `check_count_offset_target_tree_type` now captures the `open_transactional_merk_at_path` result and normalizes any open/path error to the same `Error::InvalidQuery` message so callers see a single, stable error contract for "can't run count-offset against this path." Added an explanatory comment on the function noting why we deliberately don't surface the raw open error.

File: merk/src/merk/prove.rs [hunk 2]
```diff
@@ -185,6 +189,82 @@ where
...
     }
 
+    /// Generate a sum-only proof for an `AggregateSumOnRange` query.
+    /// Mirror of [`Self::prove_aggregate_count_on_range`] for the
+    /// `ProvableSumTree` flavor.
+    ///
+    /// The merk's `tree_type` must be `ProvableSumTree`; any other tree type
+    /// is rejected with `Error::InvalidProofError` before any walking
+    /// happens. Empty merk: returns `(empty proof, sum = 0)`.
+    /// Generate an offset-paginated proof for a single-range query
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply merk/src/merk/prove.rs]: This is Claude. Fixed in b3f203d0 — removed the stale `AggregateSumOnRange` / `ProvableSumTree` rustdoc block that ended up above `prove_count_offset_on_range` after the previous insertion. The function's own docstring (starting "Generate an offset-paginated proof…") is unchanged.

File: merk/src/proofs/query/count_offset/verify.rs [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+        offset_remaining: offset,
+        limit_remaining: limit,
+        skipped: 0,
+        returned: Vec::new(),
+        left_to_right,
+    };
+    match verify_count_offset_shape(&tree, inner_range, None, None, &mut state) {
+        Ok(_struct_count) => {}
+        Err(e) => return Err(e).wrap_with_cost(cost),
+    }
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply merk/src/proofs/query/count_offset/verify.rs]: This is Claude. Fixed in b3f203d0 — switched the phase-2 verifier call to `cost_return_on_error_no_add!`. The `_no_add` variant is the right one because `verify_count_offset_shape` returns a plain `Result<u64, Error>` (it doesn't accumulate cost itself), so we need wrap-with-cost but not an add. Also added the macro to the import line.

**Comment by QuantumExplorer** [comment]
This is Claude. Pushed [`b3f203d0`](https://github.com/dashpay/grovedb/commit/b3f203d0) addressing all six CodeRabbit comments:

**Actionable (3):**
- **#1** ([`generate.rs:223`](https://github.com/dashpay/grovedb/pull/669/files#diff-7b5168bd00fc06ee1ad9bc4f1a3e3a90R223)): `check_count_offset_target_tree_type` now captures the `open_transactional_merk_at_path` result and normalizes any open/path error to `Error::InvalidQuery`, giving callers a single stable error contract.
- **#2** ([`merk/prove.rs:199`](https://github.com/dashpay/grovedb/pull/669/files#diff-25b78fef39b53fff60c10a39c1aa6cc1R199)): Removed the stale `prove_aggregate_sum_on_range` rustdoc block that ended up above `prove_count_offset_on_range`.
- **#3** ([`verify.rs:158`](https://github.com/dashpay/grovedb/pull/669/files#diff-c52ee70830a3a06f7bd14f9f9bdb6614R158)): Switched the phase-2 verifier call to `cost_return_on_error_no_add!` (the `_no_add` variant matches the plain `Result<u64, Error>` return type of `verify_count_offset_shape`).

**Nitpicks (3):**
- **#4** ([test assertions](https://github.com/dashpay/grovedb/pull/669/files#diff-92efc1717862cb87fa2cd72d11d0ec23R207)): Both rejection tests now `matches!(result, Err(Error::InvalidQuery(_)))` so unrelated errors don't accidentally satisfy them.
- **#5** ([`generate.rs:486` + `:1411`](https://github.com/dashpay/grovedb/pull/669/files#diff-7b5168bd00fc06ee1ad9bc4f1a3e3a90R486)): Both call sites now wrap with `Error::CorruptedData(format!("prove_count_offset_on_range failed: {}", e))`, matching the existing `prove_aggregate_sum_on_range` wrapping.
- **#6** ([`verify.rs:57`](https://github.com/dashpay/grovedb/pull/669/files#diff-c52ee70830a3a06f7bd14f9f9bdb6614R57)): Switched to `query.has_non_zero_offset()` for consistency with the leaf-level dispatches.

All three resolved threads were already marked resolved on the platform side. Posted one-line reply on each.

Test totals unchanged: count_offset 35/35 (17 merk + 18 grovedb), full merk 510/510, full grovedb 1731/1731.

File: grovedb/src/operations/proof/verify.rs [hunk 4]
```diff
@@ -420,6 +423,85 @@ impl GroveDb {
...
+                    current_path.iter().map(|p| p.to_vec()).collect(),
+                    proved_key_optional_value,
+                );
+                result.push(path_key_optional_value.try_into_versioned(grove_version)?);
+                limit_left
+                    .iter_mut()
+                    .for_each(|limit| *limit = limit.saturating_sub(1));
+            }
+            return Ok(count_offset_result.root_hash);
+        }
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply grovedb/src/operations/proof/verify.rs:642]: This is Claude. Fixed in 83cc4895 — both the V0 and V1 count-offset short-circuits now check `layer_proof.lower_layers.is_empty()` immediately after the syntactic gate and return `Error::InvalidProof` otherwise. Added `rejects_count_offset_proof_with_forged_lower_layers` which decodes an honest proof envelope, surgically attaches a bogus child layer to the leaf, re-encodes, and confirms the verifier rejects.

File: grovedb/src/operations/proof/verify.rs [hunk 5]
```diff
@@ -420,6 +423,85 @@ impl GroveDb {
...
+                        key: item.key.clone(),
+                        value: Some(item.value.clone()),
+                        proof: v_hash,
+                        child_hash_verified: true,
+                    };
+                let path_key_optional_value = ProvedPathKeyOptionalValue::from_proved_key_value(
+                    current_path.iter().map(|p| p.to_vec()).collect(),
+                    proved_key_optional_value,
+                );
+                result.push(path_key_optional_value.try_into_versioned(grove_version)?);
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply grovedb/src/operations/proof/verify.rs]: This is Claude. Fixed in 83cc4895. Three-part fix:

1. Extended `CountOffsetReturnedItem` with `value_hash: CryptoHash` and `child_hash_verified: bool`.
2. The merk verifier now surfaces these per-item: for `KVCount` it computes `H(value)` explicitly; for `KVValueHashFeatureType`/`KVValueHash` it forwards the proof-carried value-hash unchanged. `child_hash_verified` is always `false` (the prover never emits `KVValueHashFeatureTypeWithChildHash`).
3. The GroveDB layer uses the surfaced metadata in both V0 and V1 short-circuits.

Plus defense-in-depth: both short-circuits now deserialize each returned item and return `Error::NotSupported` if any deserializes to a non-empty tree — since the prover can't yet emit the `WithChildHash` variant, we explicitly close that soundness gap. Added `rejects_count_offset_with_non_empty_tree_return` covering it.

If the count-offset path needs to support non-empty tree returns later, the prover can be extended to emit `KVValueHashFeatureTypeWithChildHash` (mirroring the regular flow's V1 path), at which point this rejection lifts.

File: grovedb/src/query/mod.rs [hunk 6]
```diff
@@ -205,6 +205,104 @@ impl SizedQuery {
...
+    ///
+    /// Eligibility rules (all required):
+    ///
+    /// - `offset.is_some() && offset != Some(0)` — there must actually be
+    ///   an offset to honor. (Queries with offset = `None` / `Some(0)`
+    ///   take the regular proof path, which already handles them.)
+    /// - The underlying `Query` has exactly one item, and that item is a
+    ///   plain range (`Key`, `Range`, `RangeInclusive`, `RangeFrom`,
+    ///   `RangeFull`, `RangeTo`, `RangeToInclusive`, or `RangeAfter*`).
+    ///   Aggregate-count / aggregate-sum wrappers are rejected — they
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply grovedb/src/query/mod.rs]: This is Claude. Fixed in the rustdoc — removed `Key` from the plain-range list and added an explicit note that `QueryItem::Key` is rejected because a single-key match has at most one element and therefore offset > 0 is structurally guaranteed to return zero items. The rejection itself is already in place at line 293; this just brings the doc text in line with it.

File: merk/src/proofs/query/count_offset/tests.rs [hunk 7]
```diff
@@ -0,0 +1,1036 @@
...
+    merk: &Merk<impl grovedb_storage::StorageContext<'static>>,
+    expected_root: [u8; 32],
+    inner_range: QueryItem,
+    offset: u64,
+    limit: Option<u64>,
+    left_to_right: bool,
+    expected_skipped: u64,
+    expected_keys: &[&[u8]],
+    grove_version: &GroveVersion,
+) -> Vec<Vec<u8>> {
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply merk/src/proofs/query/count_offset/tests.rs:64]: This is Claude. Pushing back on this one — the analysis is incorrect. `Merk<impl grovedb_storage::StorageContext<static>>` is `impl Trait` in **argument position** (APIT), which is permitted in stable Rust. APIT desugars to an anonymous generic parameter, and it is allowed to appear nested inside a generic argument like `&Merk<...>`. What Rust does **not** permit is `impl Trait` nested inside another `impl Trait`s bound — e.g. `impl Fn(impl Trait)` — which is the case the web-search result was describing.

Evidence:
- The identical pattern is already used in `merk/src/proofs/query/aggregate_count/tests.rs:235`, `:474`, and `merk/src/proofs/query/aggregate_sum/tests.rs:102`, `:477` — all on the develop branch and compiling in CI.
- The current commit (2aa3d6e1) builds cleanly: `cargo build -p grovedb --tests` finishes without errors, and the merk test suite (529/529) passes on the same file.
- The Rust reference allows APIT in any function-parameter type position, including as a type argument of a path type: <https://doc.rust-lang.org/reference/types/impl-trait.html#anonymous-type-parameters>

No code change needed here. Leaving the helper signature as-is for consistency with the sibling `aggregate_count` / `aggregate_sum` test helpers.

**Comment by QuantumExplorer** [comment]
This is Claude. Addressed both nitpicks from the CodeRabbit review summary in commit `14ab1a6a`:

**1. `grovedb/src/tests/count_offset_paginated_tests.rs:205-387` — match on `Error::InvalidQuery` variant instead of Debug output.**
Converted the 9 flagged `format!("{:?}", err)` + `msg.contains(...)` assertions to `matches!(err, crate::Error::InvalidQuery(msg) if msg.contains(...))`. The tests now pin the exact error variant in addition to the substring, so they cannot accidentally pass if the error type ever changed but happened to keep similar textual output. (The Display-format assertions at lines 744/816/888 are intentionally left alone — those are end-to-end prover tests asserting on user-facing strings that can come from different error variants.)

**2. `merk/src/proofs/query/count_offset/tests.rs:84-95` — pin one happy-path on the full returned-item payload.**
Added a new dedicated test `returned_items_carry_full_committed_payload` that prove → encode → verifies against the 15-key fixture and asserts the **full** `CountOffsetReturnedItem` for the first returned row (key `"f"`, value `[5]`, value_hash = `H([5])`, `child_hash_verified = false`), plus a sanity loop over the other two rows checking Item-flavored value_hash + the prover invariant `child_hash_verified = false`. This pins the prover/verifier metadata contract, so a regression that silently rewrote `value`, `value_hash`, or `child_hash_verified` would now be caught.

Test totals after the change: merk 530/530 (was 529, +1 new test), grovedb 1739/1739 (same set, stronger assertions).

**Review by QuantumExplorer — COMMENTED** [review]
Reviewed
